flink-runtime/src/main/java/org/apache/flink/runtime/metrics/MetricRegistryImpl.java

```diff
@@ -144,11 +135,9 @@ public MetricRegistryImpl(MetricRegistryConfiguration config) {
 						}
 					}
 
-					Class<?> reporterClass = Class.forName(className);
-					MetricReporter reporterInstance = (MetricReporter) reporterClass.newInstance();
+					final MetricReporter reporterInstance = reporterSetup.getSupplier().get();
```

Comment:
I like the idea of not instantiating the `MetricReporters` directly in the `MetricRegistryImpl`. But do we actually need lazy instantiation our couldn't we simply pass in a `Collection<MetricReporter>` to the `MetricRegistryImpl`? I think this would further simplify the `MetricRegistryImpl`.

Reply:
In FLINK-11922 I want do add support for reporter factories, with which supported reporters would allocate resources in their constructor instead of `open`. I don't want that to occur in the configuration, hence the supplier.

Reply:
see https://github.com/zentol/flink/commit/b5feb554a965913255c3a7f3ab94430bb37d6c52#diff-65246650592087a2e5a816275f23dcc3

Reply:
Can't this happen outside of the `MetricRegistryImpl` when we create the `MetricReporter` or is there a contract that certain operations can only be when we create the `MetricRegistryImpl`? If the `MetricReporter` depend on the `MetricRegistryImpl` then I understand why we need a factory. But if not, then I think it would be easier to directly start the `MetricReporters`.